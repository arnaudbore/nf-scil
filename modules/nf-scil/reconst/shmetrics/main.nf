
process RECONST_SHMETRICS {
    tag "$meta.id"
    label 'process_single'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://scil.usherbrooke.ca/containers/scilus_2.0.2.sif':
        'scilus/scilus:2.0.2' }"

    input:
        tuple val(meta), path(sh), path(mask), path(fa), path(md)

    output:
        tuple val(meta), path("*peaks.nii.gz")          , emit: peaks, optional: true
        tuple val(meta), path("*peak_values.nii.gz")    , emit: peak_values, optional: true
        tuple val(meta), path("*peak_indices.nii.gz")   , emit: peak_indices, optional: true
        tuple val(meta), path("*afd_max.nii.gz")        , emit: afd_max, optional: true
        tuple val(meta), path("*afd_total.nii.gz")      , emit: afd_total, optional: true
        tuple val(meta), path("*afd_sum.nii.gz")        , emit: afd_sum, optional: true
        tuple val(meta), path("*nufo.nii.gz")           , emit: nufo, optional: true
        tuple val(meta), path("*ventricles_mask.nii.gz"), emit: vent_mask, optional: true
        path "versions.yml"                             , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"

    def sh_basis = task.ext.sh_basis ? "--sh_basis " + task.ext.sh_basis : ""
    def relative_threshold = task.ext.relative_threshold ? "--rt " + task.ext.relative_threshold : ""
    def fodf_metrics_a_factor = task.ext.fodf_metrics_a_factor ? task.ext.fodf_metrics_a_factor : 2.0
    def fa_threshold = task.ext.fa_threshold ? "--fa_t " + task.ext.fa_threshold : ""
    def md_threshold = task.ext.md_threshold ? "--md_t " + task.ext.md_threshold : ""
    def processes = task.ext.processes ? "--processes " + task.ext.processes : ""
    def absolute_peaks = task.ext.absolute_peaks ? "--abs_peaks_and_values" : ""
    def set_mask = mask ? "--mask $mask" : ""

    if ( task.ext.peaks ) peaks = "--peaks ${prefix}__peaks.nii.gz" else peaks = ""
    if ( task.ext.peak_values ) peak_values = "--peak_values ${prefix}__peak_values.nii.gz" else peak_values = ""
    if ( task.ext.peak_indices ) peak_indices = "--peak_indices ${prefix}__peak_indices.nii.gz" else peak_indices = ""
    if ( task.ext.afd_max ) afd_max = "--afd_max ${prefix}__afd_max.nii.gz" else afd_max = ""
    if ( task.ext.afd_total ) afd_total = "--afd_total ${prefix}__afd_total.nii.gz" else afd_total = ""
    if ( task.ext.afd_sum ) afd_sum = "--afd_sum ${prefix}__afd_sum.nii.gz" else afd_sum = ""
    if ( task.ext.nufo ) nufo = "--nufo ${prefix}__nufo.nii.gz" else nufo = ""
    if ( task.ext.ventricles_mask ) vent_mask = "--mask_output ${prefix}__ventricles_mask.nii.gz" else vent_mask = ""

    """
    export ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS=1
    export OMP_NUM_THREADS=1
    export OPENBLAS_NUM_THREADS=1

    scil_fodf_max_in_ventricles.py $sh $fa $md \
        --max_value_output ventricles_fodf_max_value.txt $sh_basis \
        $fa_threshold $md_threshold $vent_mask -f

        echo "Maximal peak value in ventricle in file : \$(cat ventricles_fodf_max_value.txt)"

        a_factor=$fodf_metrics_a_factor
        v_max=\$(sed -E 's/([+-]?[0-9.]+)[eE]\\+?(-?)([0-9]+)/(\\1*10^\\2\\3)/g' <<<"\$(cat ventricles_fodf_max_value.txt)")

        echo "Maximal peak value in ventricles : \${v_max}"

        a_threshold=\$(echo "scale=10; \${a_factor} * \${v_max}" | bc)
        if (( \$(echo "\${a_threshold} <= 0" | bc -l) )); then
            a_threshold=1E-10
        fi

        echo "Computing fodf metrics with absolute threshold : \${a_threshold}"

        scil_fodf_metrics.py $sh \
            $set_mask $sh_basis $absolute_peaks \
            $peaks $peak_values $peak_indices \
            $afd_max $afd_total \
            $afd_sum $nufo \
            $relative_threshold --not_all --at \${a_threshold}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: 2.0.2
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    scil_fodf_max_in_ventricles.py -h
    scil_fodf_metrics.py -h

    touch ${prefix}__peaks.nii.gz
    touch ${prefix}__peak_values.nii.gz
    touch ${prefix}__peak_indices.nii.gz
    touch ${prefix}__afd_max.nii.gz
    touch ${prefix}__afd_total.nii.gz
    touch ${prefix}__afd_sum.nii.gz
    touch ${prefix}__nufo.nii.gz
    touch ${prefix}__ventricles_mask.nii.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: 2.0.2
    END_VERSIONS
    """
}
