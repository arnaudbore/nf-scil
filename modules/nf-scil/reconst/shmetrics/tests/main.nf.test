nextflow_process {

    name "Test Process RECONST_SHMETRICS"
    script "../main.nf"
    process "RECONST_SHMETRICS"

    tag "modules"
    tag "modules_nfcore"
    tag "reconst"
    tag "reconst/shmetrics"

    tag "subworkflows"
    tag "subworkflows/load_test_data"

    setup {
            run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
                script "../../../../../subworkflows/nf-scil/load_test_data/main.nf"
                process {
                    """
                    input[0] = Channel.from( [ "processing.zip" ] )
                    input[1] = "test.load-test-data"
                    """
                }
            }
        }

    test("reconst - shmetrics") {

        config "./nextflow.config"

        when {
            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory.map{
                    test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/fodf.nii.gz"),
                        [],
                        file("\${test_data_directory}/fa.nii.gz"),
                        file("\${test_data_directory}/md.nii.gz")
                    ]
                }
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
