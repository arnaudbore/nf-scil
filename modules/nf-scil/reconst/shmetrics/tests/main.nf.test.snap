{
    "reconst - shmetrics": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__peaks.nii.gz:md5,e06a643586eb13034d9350dd1d73618e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__peak_values.nii.gz:md5,9e325f94d68db5913a5866be39d5b8be"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__peak_indices.nii.gz:md5,82217067e7cc285857a20a762b345566"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__afd_max.nii.gz:md5,7cadad1e2ab2e04c13c64d036d2feca1"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__afd_total.nii.gz:md5,024a8c657e77c3611c5a15a9868e308d"
                    ]
                ],
                "5": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__afd_sum.nii.gz:md5,7b565fa2c005dcef9e0c6ad6af8abef0"
                    ]
                ],
                "6": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__nufo.nii.gz:md5,0ba1886fb00013fa225e2e5f12a30ab8"
                    ]
                ],
                "7": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__ventricles_mask.nii.gz:md5,bd3b550c4470eb62fe279c28879b6b5d"
                    ]
                ],
                "8": [
                    "versions.yml:md5,a07e324c55b4310c93cc4003bee2cd6c"
                ],
                "afd_max": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__afd_max.nii.gz:md5,7cadad1e2ab2e04c13c64d036d2feca1"
                    ]
                ],
                "afd_sum": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__afd_sum.nii.gz:md5,7b565fa2c005dcef9e0c6ad6af8abef0"
                    ]
                ],
                "afd_total": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__afd_total.nii.gz:md5,024a8c657e77c3611c5a15a9868e308d"
                    ]
                ],
                "nufo": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__nufo.nii.gz:md5,0ba1886fb00013fa225e2e5f12a30ab8"
                    ]
                ],
                "peak_indices": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__peak_indices.nii.gz:md5,82217067e7cc285857a20a762b345566"
                    ]
                ],
                "peak_values": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__peak_values.nii.gz:md5,9e325f94d68db5913a5866be39d5b8be"
                    ]
                ],
                "peaks": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__peaks.nii.gz:md5,e06a643586eb13034d9350dd1d73618e"
                    ]
                ],
                "vent_mask": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__ventricles_mask.nii.gz:md5,bd3b550c4470eb62fe279c28879b6b5d"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,a07e324c55b4310c93cc4003bee2cd6c"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0-rc1",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-06-19T21:53:27.213837"
    }
}